rust/connlib/tunnel/src/tests.rs

```diff
@@ -74,103 +108,659 @@ impl StateMachineTest for TunnelTest {
+                client,
+                gateway,
+                relay,
+                client_cidr_resources: IpNetworkTable::new(),
+                connected_resources: Default::default(),
+                gateway_received_icmp_packets: Default::default(),
+            })
+            .boxed()
+    }
 
-        // TODO: Drain `poll_transmit` and execute it.
+    /// Defines the [`Strategy`] on how we can [transition](Transition) from the current [`ReferenceState`].
+    ///
+    /// This is invoked by proptest repeatedly to explore further state transitions.
+    /// Here, we should only generate [`Transition`]s that make sense for the current state.
+    fn transitions(state: &Self::State) -> proptest::prelude::BoxedStrategy<Self::Transition> {
+        let add_cidr_resource = cidr_resource(8).prop_map(Transition::AddCidrResource);
+        let tick = (0..=1000u64).prop_map(|millis| Transition::Tick { millis });
+
+        let (num_ip4_resources, num_ip6_resources) = state.client_cidr_resources.len();
+
+        let weight_ip4 = if num_ip4_resources == 0 { 0 } else { 3 };
+        let weight_ip6 = if num_ip6_resources == 0 { 0 } else { 3 };
+
+        // Note: We use weighted strategies here to conditionally only include the ICMP strategies if we have a resource.
+        prop_oneof![
+            1 => add_cidr_resource,
+            1 => tick,
+            1 => icmp_to_random_ip(),
+            weight_ip4 => icmp_to_ipv4_cidr_resource(),
+            weight_ip6 => icmp_to_ipv6_cidr_resource()
+        ]
+        .boxed()
+    }
+
+    /// Apply the transition to our reference state.
+    ///
+    /// Here is where we implement the "expected" logic.
+    fn apply(mut state: Self::State, transition: &Self::Transition) -> Self::State {
+        match transition {
+            Transition::AddCidrResource(r) => {
+                state.client_cidr_resources.insert(r.address, r.clone());
+            }
+            Transition::SendICMPPacketToRandomIp { dst } => {
+                state.on_icmp_packet(state.client.tunnel_ip(*dst), *dst);
+            }
+            Transition::SendICMPPacketToIp4Resource { r_idx } => {
+                let dst = state.sample_ipv4_cidr_resource_dst(r_idx);
+                state.on_icmp_packet(state.client.tunnel_ip4, dst);
+            }
+            Transition::SendICMPPacketToIp6Resource { r_idx } => {
+                let dst = state.sample_ipv6_cidr_resource_dst(r_idx);
+                state.on_icmp_packet(state.client.tunnel_ip6, dst);
+            }
+            Transition::Tick { millis } => state.now += Duration::from_millis(*millis),
+        };
 
         state
     }
+
+    /// Any additional checks on whether a particular [`Transition`] can be applied to a certain state.
+    fn preconditions(state: &Self::State, transition: &Self::Transition) -> bool {
+        match transition {
+            Transition::AddCidrResource(r) => {
+                // TODO: PRODUCTION CODE DOES NOT HANDLE THIS!
```

Comment:
Does not handle a gateway getting a resource that it doesn't have connectivity for?